components/style/properties/longhand/box.mako.rs

```diff
@@ -13,229 +13,41 @@
-            where W: ::std::fmt::Write,
-        {
-            match *self {
-                % for value in values:
-                    SpecifiedValue::${to_camel_case(value)} => dest.write_str("${value}"),
-                % endfor
-            }
-        }
-    }
-
-    /// The initial display value.
-    #[inline]
-    pub fn get_initial_value() -> computed_value::T {
-        computed_value::T::${to_camel_case(values[0])}
-    }
-
-    /// Parse a display value.
-    pub fn parse<'i, 't>(_context: &ParserContext, input: &mut Parser<'i, 't>)
-                         -> Result<SpecifiedValue, ParseError<'i>> {
-        try_match_ident_ignore_ascii_case! { input,
-            % for value in values:
-                "${value}" => {
-                    Ok(computed_value::T::${to_camel_case(value)})
-                },
-            % endfor
-            % for value in webkit_prefixed_values:
-                "-webkit-${value}" => {
-                    Ok(computed_value::T::${to_camel_case(value)})
-                },
-            % endfor
-        }
-    }
-
-    % if product == "servo":
-        fn cascade_property_custom(_declaration: &PropertyDeclaration,
-                                   context: &mut computed::Context) {
-            longhands::_servo_display_for_hypothetical_box::derive_from_display(context);
-            longhands::_servo_text_decorations_in_effect::derive_from_display(context);
-        }
-    % endif
+${helpers.predefined_type(
+    "display",
+    "Display",
+    "computed::Display::inline()",
+    initial_specified_value="specified::Display::inline()",
+    animation_value_type="discrete",
+    needs_context=False,
+    custom_cascade= product == 'servo',
+    custom_cascade_function="specified::Display::cascade_property_custom",
+    flags="APPLIES_TO_PLACEHOLDER",
+    spec="https://drafts.csswg.org/css-display/#propdef-display",
+)}
 
-    ${helpers.gecko_keyword_conversion(Keyword('display', ' '.join(values),
-                                               gecko_enum_prefix='StyleDisplay',
-                                               gecko_strip_moz_prefix=False))}
+<%
+    display_values = """inline block inline-block
+    table inline-table table-row-group table-header-group table-footer-group
+    table-row table-column-group table-column table-cell table-caption
+    list-item none
+    """.split()
+    webkit_prefixed_display_values = "flex inline-flex".split()
+    display_values += webkit_prefixed_display_values
+    if product == "gecko":
+        display_values += """grid inline-grid ruby ruby-base ruby-base-container
+            ruby-text ruby-text-container contents flow-root -webkit-box
+            -webkit-inline-box -moz-box -moz-inline-box -moz-grid -moz-inline-grid
+            -moz-grid-group -moz-grid-line -moz-stack -moz-inline-stack -moz-deck
+            -moz-popup -moz-groupbox""".split()
+%>
 
-</%helpers:longhand>
+${helpers.gecko_keyword_conversion(
```

Comment:
It's very unfortunate that we still need to list all the values here... Can you double-check that adding a new value to the `Display` enum at least causes a compile error here?

And can you leave a `FIXME` above like:

```python
# FIXME(emilio): Listing all the display values here is very unfortunate, we should teach C++ to use the
# Rust enum directly, or generate the conversions to `StyleDisplay`.
```

Reply:
Sure, I'll double-check the `Display` enum again and add this `FIXME` comment here.

Reply:
Yeah, I tried to add a `display-test` into `display_values` and it caused compilation error with 

```diff
diff --git a/components/style/properties/longhand/box.mako.rs b/components/style/properties/longhand/box.mako.rs
index 641d72afae..b32248b513 100644
--- a/components/style/properties/longhand/box.mako.rs
+++ b/components/style/properties/longhand/box.mako.rs
@@ -32,7 +32,7 @@ ${helpers.predefined_type(
     display_values = """inline block inline-block
     table inline-table table-row-group table-header-group table-footer-group
     table-row table-column-group table-column table-cell table-caption
-    list-item none
+    list-item none display-test
     """.split()
     webkit_prefixed_display_values = "flex inline-flex".split()
     display_values += webkit_prefixed_display_values
```

```rust
   Compiling style v0.0.1 (file:///Users/cybai/codespace/servo/components/style)
error[E0599]: no associated item named `DisplayTest` found for type `gecko_bindings::structs::root::mozilla::StyleDisplay` in the current scope
    --> servo/target/geckolib/debug/build/style-55046d2555925070/out/properties.rs:7989:23
     |
7989 |                     = structs::StyleDisplay::DisplayTest as u32;
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `style`.

To learn more, run the command again with --verbose.
GeckoLib build completed in 0:00:18
```